### PR TITLE
Ehentai dev

### LINKED ISSUE #7665: 【Rss Wanted】E站
#### 主页

https://e-hentai.org/

#### 收藏

https://e-hentai.org/favorites.php?favcat=0

https://e-hentai.org/favorites.php?page=1

https://e-hentai.org/favorites.php?inline_set=fs_p

#### 标签

https://e-hentai.org/tag/male:glasses

https://e-hentai.org/tag/parody:original

#### 搜索

https://e-hentai.org/?f_cats=256&f_search=parody%3Aoriginal%24

https://e-hentai.org/?f_cats=384&f_search=parody%3Aoriginal%24

### 网站描述

E站

### 需要生成什么内容？

标题、作者、发布时间、链接

对于收藏，还需要添加收藏的时间

### PR DESCRIPTION
**Pull request by yindaheng98**
## 该 PR 相关 Issue / Involved issue

Close #7665 

## 完整路由地址 / Example for the proposed route(s)

```routes
/ehentai/favorites/0/posted/1
/ehentai/tag/language:chinese
/ehentai/search/page=3&advsearch=1&f_sdesc=on&f_storr=on&f_sto=on&f_spf=&f_spt=/1
```

## 新RSS检查列表 / New RSS Script Checklist

- [x] 这个PR中包含了新的路由吗? Does this PR add new route?
  - 如果有, 请完成检查列表. If yes, please finish the check list
  - **如果你的PR符合下方某个事项, 也请注明. If any of the checklist item meets your PR, please fill it out.**
  - [x] <- 这样打勾
- [x] 是否提供了文档? Documentation provided?
  - [ ] 是否提供了英文文档? EN Documentation provided?
- [ ] 是否支持全文获取? Is this RSS Script support fulltext?
  - [ ] 如果全文获取中需要访问文章链接, 是否使用了缓存? If fulltext requires to fetch detail pages, is cache used in the process?
  - [缓存说明](https://docs.rsshub.app/joinus/quick-start.html#ti-jiao-xin-de-rsshub-gui-ze-bian-xie-jiao-ben-shi-yong-huan-cun) | [How to use cache](https://docs.rsshub.app/joinus/quick-start.html#ti-jiao-xin-de-rsshub-gui-ze-bian-xie-jiao-ben-shi-yong-huan-cun)
- [ ] 目标是否有明显的反爬/频率限制? Is there any sign of anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? (延长缓存时间, 写文档说明, etc.) If yes, do your code reflect this sign? (e.g. write documentations, use long cache time)
- [x] 目标是否有提供日期？ Is there a date in the source?
  - [x] 如果有，包是否正确解析？ If there is, can this script provide this info?
  - [x] 如果有提供解析能力，时区是否正确调整？ Is the timezone correctly provided?
  - 如果有提供日期，但是没有提供解析，请说明原因 If there is a date but this script does not parse, please provide your reason.
- [ ] 是否引入的新的包? Any new package introduced?
  - 如果有, 请说明原因. If yes, please state your reason
- [ ] 是否使用了`Puppeteer`? Make use of `Puppeteer`?
  - 如果有, 请说明原因. If yes, please state your reason
  

## 说明 / Note

- 经测试，E站反爬限制似乎是基于Image Limits的
- 开启`EH_GET_BITTORRENT`爬取种子文件，一次大概5 Image Limits

### CODE REVIEW

File: lib/routes/ehentai/ehapi.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+function full_url(url) {
+    return `https://e-hentai.org/${url}`;
+}
+
+const headers = { Host: 'e-hentai.org' };
+const has_cookie = config.ipb_member_id && config.ipb_pass_hash && config.ipd_session_id && config.sk;
+if (has_cookie) {
+    const ipb_member_id = config.ipb_member_id;
+    const ipb_pass_hash = config.ipb_pass_hash;
+    const ipd_session_id = config.ipd_session_id;
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:14]: Not required.

Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:14]: Not required.

File: lib/routes/ehentai/ehapi.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+function generalGot(url) {
+    return got({
+        method: 'get',
+        url: url,
+        headers: headers,
+    });
+}
+
+function ehgot(url) {
+    return generalGot(full_url(url));
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:28]:
```suggestion
    return generalGot(`https://e-hentai.org/${url}`);
```

File: lib/routes/ehentai/ehapi.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const config = require('@/config').value.ehentai;
+
+function full_url(url) {
+    return `https://e-hentai.org/${url}`;
+}
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:7]:
```suggestion
```

**Review by SukkaW — CHANGES_REQUESTED** [review]
Please add built-in cache support. E-Hentai has a very strict rate limit, any aggressive requesting will be considered as violating the rules, and will *always* result in an account ban for 1000 years (a.k.a. being sent to the moon).

File: lib/routes/ehentai/ehapi.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+const headers = { Host: 'e-hentai.org' };
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:9]: Really? Why specifying `Host`?

**Review by yindaheng98 — COMMENTED** [review]
reviewed

File: lib/routes/ehentai/ehapi.js (hunk 1 above)
Comment (yindaheng98) [reply lib/routes/ehentai/ehapi.js]: 已删除

File: lib/routes/ehentai/ehapi.js (hunk 4 above)
Comment (yindaheng98) [reply lib/routes/ehentai/ehapi.js]: 确实不需要`Host`字段，已删除

File: docs/picture.md [hunk 5]
```diff
@@ -477,3 +477,17 @@ R18 显示
 ### 最新主题
 
 <Route author="nczitzk" example="/zhutix/latest" path="/zhutix/latest"/>
+
+## eHentai
```
Comment (SettingDust) [inline docs/picture.md:481]: 放到二次元里好些吧？主体还是漫画

**Review by SettingDust — APPROVED** [review]

File: docs/picture.md (hunk 5 above)
Comment (yindaheng98) [reply docs/picture.md:481]: 这我感觉都一样，不影响
我是看nhentai在picture分类里面所以就把e站放进来的

**Comment by NeverBehave** [comment]
我想暂缓一下这个pr的合并，在cache动议完成以后
因为这个路由可能使用的人数比较多

**Comment by yindaheng98** [comment]
> Please add built-in cache support. E-Hentai has a very strict rate limit, any aggressive requesting will be considered as violating the rules, and will _always_ result in an account ban for 1000 years (a.k.a. being sent to the moon).

已参考bilibili路由添加了获取种子文件链接的缓存方案

File: lib/router.js [hunk 6]
```diff
@@ -4140,6 +4140,11 @@ router.get('/tongli/news/:type', lazyloadRouteHandler('./routes/tongli/news'));
 // OR
 router.get('/or/:id?', lazyloadRouteHandler('./routes/or'));
 
+// e-hentai
+router.get('/ehentai/favorites/:favcat?/:order?/:page?/:bittorrent?', require('./routes/ehentai/favorites'));
+router.get('/ehentai/search/:params?/:page?/:bittorrent?', require('./routes/ehentai/search'));
+router.get('/ehentai/tag/:tag/:page?/:bittorrent?', require('./routes/ehentai/tag'));
```
Comment (SukkaW) [inline lib/router.js:4146]:
```suggestion
// e-hentai
router.get('/ehentai/favorites/:favcat?/:order?/:page?/:bittorrent?', lazyloadRouteHandler('./routes/ehentai/favorites'));
router.get('/ehentai/search/:params?/:page?/:bittorrent?', lazyloadRouteHandler('./routes/ehentai/search'));
router.get('/ehentai/tag/:tag/:page?/:bittorrent?', lazyloadRouteHandler('./routes/ehentai/tag'));
```

File: lib/routes/ehentai/ehapi.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const config = require('@/config').value.ehentai;
+
+const headers = {};
+const has_cookie = config.ipb_member_id && config.ipb_pass_hash && config.sk;
+if (has_cookie) {
+    const ipb_member_id = config.ipb_member_id;
+    const ipb_pass_hash = config.ipb_pass_hash;
+    const sk = config.sk;
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:10]:
```suggestion
    const { ipb_member_id, ipb_pass_hash, sk } = config;
```

File: docs/picture.md [hunk 8]
```diff
@@ -479,3 +479,17 @@ R18 显示
...
```
Comment (SukkaW) [inline docs/picture.md:483]:
```suggestion
## E-Hentai
```

File: docs/install/README.md [hunk 9]
```diff
@@ -702,3 +702,8 @@ RSSHub 支持使用访问密钥 / 码，白名单和黑名单三种方式进行
         | <https://pic1.xuehuaimg.com/proxy/>      | cloudflare   |
         | <https://cors.netnr.workers.dev/>        | cloudflare   |
         | <https://netnr-proxy.openode.io/>        | digitalocean |
+
+-   E-Hentai
+    -   `EH_IPB_MEMBER_ID`: E站浏览器cookie中的`ipb_member_id`值
+    -   `EH_IPB_PASS_HASH`: E站浏览器cookie中的`ipb_pass_hash`值
+    -   `EH_SK`: E站浏览器cookie中的`sk`值
```
Comment (SukkaW) [inline docs/install/README.md:709]:
```suggestion
    -   `EH_IPB_MEMBER_ID`: E-Hentai 账户登录后 cookie 的 `ipb_member_id` 值
    -   `EH_IPB_PASS_HASH`: E-Hentai 账户登录后 cookie 的 `ipb_pass_hash` 值
    -   `EH_SK`: E-Hentai 账户登录后 cookie 中的`sk`值
```

File: lib/routes/ehentai/ehapi.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        url: url,
+        headers: headers,
+    });
+}
+
+function ehgot(url) {
+    return generalGot(`https://e-hentai.org/${url}`);
+}
+
+ehgot.has_cookie = has_cookie;
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:26]: What does this line do?

File: lib/routes/ehentai/ehapi.js [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    headers.cookie = `ipb_member_id=${ipb_member_id};ipb_pass_hash=${ipb_pass_hash};sk=${sk}`;
+}
+
+function generalGot(url) {
+    return got({
+        method: 'get',
+        url: url,
+        headers: headers,
+    });
+}
```
Comment (SukkaW) [inline lib/routes/ehentai/ehapi.js:20]: Isn't the `got.get` not good enough? Why wrap it in a separate function?

**Review by SukkaW — CHANGES_REQUESTED** [review]

File: lib/routes/ehentai/ehapi.js (hunk 10 above)
Comment (yindaheng98) [reply lib/routes/ehentai/ehapi.js:16]: It is useless, deleted.

File: lib/routes/ehentai/ehapi.js (hunk 11 above)
Comment (yindaheng98) [reply lib/routes/ehentai/ehapi.js]: You right, it is good enough, just use it. Related code modified.

**Review by SukkaW — APPROVED** [review]

**Pull request merged**